### DISCUSSION TITLE
node flags must be passed to `node`, not `ts-node`

### DISCUSSION THREAD

**Discussion by cspotcode** [TypeStrong/ts-node#1182] [Official Recipes] [answered]
If you want to use various [`node` flags](https://nodejs.org/dist/latest-v15.x/docs/api/cli.html) with `ts-node`, you must pass them to `node` directly.  For example, here we are loading `ts-node` and also passing node's `-gc` flag:

```
node -r ts-node/register -gc ./index.ts
```

Node also supports a [`NODE_OPTIONS` environment variable, explained in node's official documentation](https://nodejs.org/api/cli.html#cli_node_options_options).  You can use that environment variable with `ts-node`.

```
NODE_OPTIONS="-gc" ts-node-script ./index.ts
```

If you're curious why we do this, you can read this issue, which explains why spawning a subprocess causes other problems.  #471

Remember that the examples above are written in bash syntax.  If you're using Windows PowerShell or cmd, they have buggy, confusing, and/or limiting behavior for parsing a command-line and setting environment variables.  `ts-node` uses node's standard `process.argv` and `process.env`.  We are not responsible for how they are parsed or passed into the node process.  If a Windows shell is not giving us the correct `process.argv` or `process.env` values, then this is the fault of the Windows shell.

To debug the values being passed to `ts-node`, try this:

```
ts-node --transpile-only -p -e "console.dir(process.argv)"
ts-node --transpile-only -p -e "console.dir(process.execArgv)"
ts-node --transpile-only -p -e "console.dir(process.env)"

node -p "console.dir(process.argv)"
node -p "console.dir(process.execArgv)"
node -p "console.dir(process.env)"
```

Remember also that `npm` uses a cmd shim for all installed binaries.  This means when you run `ts-node`, you are invoking a `cmd.exe` batch script, which invokes `node`, passing it the path to `ts-node`'s entrypoint .js file.  This behavior is controlled by `npm`, and there are certain limitations and quirks imposed by this cmd shim.

In the debugging examples above, the first 3 are using this cmd shim.  The last 3 are not.  If you see discrepancies, it is caused by the limitations and quirks of npm's cmd shim.

**Comment by cspotcode** [comment, marked as answer]
If the information above can be improved in any way, please ask questions here and we will update it accordingly.

**Comment by Glench** [comment]
Thanks for sharing! Setting the `NODE_OPTIONS` environment variable was really helpful to pass the `-C` flag to node.
